cot/src/common_types.rs

```diff
@@ -0,0 +1,474 @@
+    }
+
+    /// Returns the full email address as a string.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::str::FromStr;
+    ///
+    /// use cot::common_types::Email;
+    ///
+    /// let email = Email::from_str("user@example.com").unwrap();
+    /// assert_eq!(email.email(), "user@example.com");
+    /// ```
+    #[must_use]
+    pub fn email(&self) -> String {
+        self.0.email()
+    }
+
+    /// Returns the local part of the email address (the part before the '@'
+    /// symbol).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::str::FromStr;
+    ///
+    /// use cot::common_types::Email;
+    ///
+    /// let email = Email::from_str("user@example.com").unwrap();
+    /// assert_eq!(email.local_part(), "user");
+    /// ```
+    #[must_use]
+    pub fn local_part(&self) -> &str {
+        self.0.local_part()
+    }
+
+    /// Returns the display part of the email address.
+    ///
+    /// For simple email addresses, this is typically the same as the local
+    /// part. For email addresses with display names, this returns the
+    /// display name portion.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::str::FromStr;
+    ///
+    /// use cot::common_types::Email;
+    ///
+    /// let email = Email::from_str("Name <name@example.org>").unwrap();
+    /// assert_eq!(email.display_part(), "Name".to_owned());
+    /// ```
+    #[must_use]
+    pub fn display_part(&self) -> &str {
+        self.0.display_part()
+    }
+}
+
+/// Implements string parsing for `Email`.
+///
+/// # Examples
+///
+/// ```
+/// use std::str::FromStr;
+///
+/// use cot::common_types::Email;
+///
+/// let email = Email::from_str("user@example.com").unwrap();
+/// ```
+impl FromStr for Email {
+    type Err = email_address::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Email::new(s)
+    }
+}
+
+/// Implements conversion from string references to `Email`.
+///
+/// # Examples
+///
+/// ```
+/// use cot::common_types::Email;
+///
+/// let email = Email::try_from("user@example.com").unwrap();
+/// ```
+impl TryFrom<&str> for Email {
+    type Error = email_address::Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Email::new(value)
+    }
+}
+
+/// Implements conversion from `String` to `Email`.
+///
+/// # Examples
+///
+/// ```
+/// use cot::common_types::Email;
+///
+/// let email = Email::try_from(String::from("user@example.com")).unwrap();
+/// ```
+#[cfg(feature = "db")]
+impl TryFrom<String> for Email {
+    type Error = email_address::Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Email::new(value)
+    }
+}
+
+/// Implements database value conversion for `Email`.
+///
+/// This allows `Email` to be stored in the database as a text value.
+#[cfg(feature = "db")]
+impl ToDbValue for Email {
+    fn to_db_value(&self) -> DbValue {
+        self.0.clone().to_string().into()
```

Comment:
This is unstable because `EmailAddress` returns the inner `String` in its `Display` [implementation](https://docs.rs/email_address/latest/src/email_address/lib.rs.html#595-597), which is used in the `to_string()` call. So, if a user provides `Name <name@example.org>` one time and `Name <name@example.org>` the other time, this will return different values despite being logically the same email. If `Email` is used just as an info field, this wouldn't matter that much, but it could be used for a key, and then it's crucial.

This should fix it:
```suggestion
        self.0.clone().email().into()
```

I'd like to see a test for that situation as well, please.

Reply:
ah yeah, nice catch, thanks for pointing that out.